sdk/burdock/query/sql/ast/tokens.py

```diff
@@ -188,6 +188,12 @@ def evaluate(self, bindings):
         raise ValueError("We don't know how to evaluate " + str(self))
     def children(self):
         return [None]
+    def symbol(self, relations):
```

Comment:
Where is this new method used?

Reply:
This method was always part of the code, but I forgot to override it on some of the AST nodes.  The behavior of symbol() and load_symbols() is specified in the dev spec at https://drive.google.com/file/d/1XHZUiYpYcMP6Ix56lq-Og6iU33GI8cqV/view.  The "AST" heading gets into load_symbols, which is what resolves curator-supplied symbols, and the "Expressions" heading talks about symbol() which is what resolves analyst-provided symbols.  Those two methods basically connect analyst and curator symbols, since both sources are externally-provided and can have various inconsistencies.

The bug here was that I implemented, symbol() on the base classes to just pass through.  That was great for AST nodes that don't actually reference any database columns.  However, some AST nodes reference database columns, maybe nested way down the tree, and passing them through without recursively walking was leaving the database column symbols unresolved.

On the lowest base class, I now only pass through if the node doesn't reference database columns.  If it references database columns, it means we should've overriden the method on the specialized class.  So I throw in that case.  I also added this case on bare "Expression", because the bare expression is a pass-through that we can't even check to see if it has children.  We should almost never actually use this class anyway; it's like an escape valve for when we haven't implemented a more specialized expression in the parser.  Since we can't verify if it's safe to pass through the symbol, we throw here.  Both exceptions should never be hit.

I also went through and added the symbol everywhere I forgot it.  It's worth doing another pass to make sure we didn't miss any.  Anywhere that might reference a database column needs to have a .symbol() attached.  For example, people might use a database column to set the exponent on a power, even though it's a literal 99.9% of the time, we still need to try to resolve it.

This also raises another point about about how we do these type of operations in the code.  There are at least 3 places like this where we are traversing the tree and doing things based on the values.  My current pattern in all these cases is to use recursion.  That makes it important to put in safeguards like the exceptions I added here, in case we forget to implement the recursion methods on an inherited class.

In Scala, I would probably instead do a deep copy with pattern matching to swap out the changed nodes as we copy.  We could hack together something like that for Python, but it doesn't feel right.  Maybe there are benefits to that approach though.

Another way to do it would be like what we could do in C# or Java, where we locate the target nodes in the tree using an xpath-style accessor, (e.g. .findnodes(Column)) and then use the references to modify them in-place.  I implemented the two tree-accessor functions and use them in multiple places, but Python doesn't seem to like in-place editing of objects returned that way.  Or maybe I just don't understand how to return references rather than copies?  Anyway, this pattern has some advantages and disadvantages, but we could maybe make that work.

I'm OK with sticking with recursion as well; just open to rethinking it.